src/code/z_actor.c

```diff
@@ -3646,187 +3646,286 @@ Hilite* func_8003435C(Vec3f* object, PlayState* play) {
     return func_8002EB44(object, &play->view.eye, &lightDir, play->state.gfxCtx);
 }
 
-s32 func_800343CC(PlayState* play, Actor* actor, s16* arg2, f32 interactRange, callback1_800343CC unkFunc1,
-                  callback2_800343CC unkFunc2) {
+/**
+ * Updates NPC talking state. Checks for a talk request and updates
+ * the talkState parameter when a dialog is ongoing. Otherwise checks if
+ * the actor is onscreen, advertises the interaction in a range and sets
+ * the current text id if necessary.
+ *
+ * The talk state values are defined in the NpcTalkState enum.
+ *
+ * @see NpcTalkState
+ *
+ * @param[in,out] talkState Talk state
+ * @param interactRange The interact (talking) range for the actor
+ * @param getTextId Callback for getting the next text id
+ * @param updateTalkState Callback for getting the next talkState value
+ * @return True if a new dialog was started (player talked to the actor). False otherwise.
+ */
+s32 Npc_UpdateTalking(PlayState* play, Actor* actor, s16* talkState, f32 interactRange, NpcGetTextIdFunc getTextId,
+                      NpcUpdateTalkStateFunc updateTalkState) {
     s16 x;
     s16 y;
 
     if (Actor_ProcessTalkRequest(actor, play)) {
-        *arg2 = 1;
+        *talkState = NPC_TALK_STATE_TALKING;
         return true;
     }
 
-    if (*arg2 != 0) {
-        *arg2 = unkFunc2(play, actor);
+    if (*talkState != NPC_TALK_STATE_IDLE) {
+        *talkState = updateTalkState(play, actor);
         return false;
     }
 
     Actor_GetScreenPos(play, actor, &x, &y);
-
     if ((x < 0) || (x > SCREEN_WIDTH) || (y < 0) || (y > SCREEN_HEIGHT)) {
+        // Actor is offscreen
         return false;
     }
 
     if (!func_8002F2CC(actor, play, interactRange)) {
+        // Player cannot talk to actor now
         return false;
     }
```

Comment:
I think this is less so "the player cant talk" and more "the player hasnt started the conversation yet"
I think it may be good to just omit a comment until we decide on how we will phrase this conversation-starting function names (and after that point, a comment will probably be redundant anyway)

Reply:
I'm pretty sure `func_8002F2CC` returns true when the A button shows "Speak" (player is in talk range) and does not actually start the conversation. But the point stands; I will remove the comment 👍🏼

Reply:
it shouldnt cause of that first check for `ACTOR_FLAG_8`
from my actor flag docs branch:
`#define ACTOR_FLAG_TALK_REQUESTED (1 << 8) // Player has requested to talk to the actor; Player uses this flag differently than every other actor`

but yea, sounds good

Reply:
If ACTOR_FLAG_8 is set the function just returns false right away, right?
```c
    if ((player->actor.flags & ACTOR_FLAG_8) || 
        ((exchangeItemId != EXCH_ITEM_NONE) && Player_InCsMode(play)) ||
        (!actor->isTargeted &&
         ((arg3 < fabsf(actor->yDistToPlayer)) || (player->targetActorDistance < actor->xzDistToPlayer) ||
          (arg2 < actor->xzDistToPlayer)))) {
        return false;
    }
 ```
 
This is why I've been adding the comments like this one, just so the logic is a bit clearer to read until the functions are renamed 😅

Reply:
Not to start discussing actor flags, but note `player->actor.flags & ACTOR_FLAG_8` is a check on the **player** actor's flags